Pitch filter improvement

jclsn opened the pull request:
Adds a simple Kalman filter with a constant velocity model as alternative pitch estimator. The model assumes constant playback, which is realized by assuming that every zero-crossing corresponds to a quarter revolution of the sinusoid. The quarter revolution is fed as displacement into the filter and the new velocity (pitch) is estimated.

The results are much better than with the Alpha-Beta filter, yet not perfect. It would probably be best to take the time to statistically model the filter better. For now definitely an improvement though!

Adds a selection in the vinylcontrol panel to choose between pitch estimators for people who don't trust the new filter just yet.

jclsn committed:
xwax: Implement a Kalman filter as alternative pitch estimator

This implements a Kalman filter with mode switches for different
responsiveness. The results are much better than with the Alpha-Beta
filter. Still not perfect. It would probably be best to model a
continuous curve for this.

For now definitely an improvement though!

Swiftb0y commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,157 @@
 
-    p->x -= dx; /* relative to previous */
+/* Optionally change dt (if your cadence changes) */
+static inline void pitch_set_dt(struct pitch *p, double dt)
+{
+    p->dt = dt;
 }
```
this seems to be unused. I guess this is supposed to be called on samplerate changes so the filter adapts without having to reinitialize the entire timecoder?

jclsn replied:
Yes, usually `dt is also the wrong name in discrete time. This is just inherited from xwax. Usually `Ts = 1 / Fs` is the time step in this discrete domain `dt` for the continuous domain.

jclsn replied:
Removed

Swiftb0y commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -614,14 +614,13 @@ void VinylControlXwax::analyzeSamples(CSAMPLE* pSamples, size_t nFrames) {
-                // converge on the correct amount.
-                m_dDisplayPitch += pitch_difference * .25;
-            } else {
+            }  else {
                 // For extremely small changes, converge very slowly.
                 m_dDisplayPitch += pitch_difference * .01;
             }
+
+            m_dDisplayPitch = dVinylPitch;
+
```
This needs clean up. The entire `if` above is now dead code.

Swiftb0y reviewed (commented):
couple thoughts. Not an expert on Kalman filters, but as far as DSP code goes, I've seen way uglier.

jclsn commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,157 @@
 
-/* Get the pitch after filtering */
+/* Feed one observation: in the last dt seconds, position moved by dx */
+static inline void pitch_dt_observation(struct pitch *p, double dx)
```
Isn't the plan to go for vinylla?

Swiftb0y replied:
> alas, xwax is pure C. I could see an argument for creating our own C++ fork / upgrade of libxwax, but I would also want Mark's blessing to do so since that would mean a permanent divergence.

Yeah, wasn't a serious suggestion. Just dreaming of a "better tomorrow" ;)

ywwg commented:
Instead of replacing the existing filters, would it be possible to make this a new set of functions, and xwax could be configured to use one or the other at compile time?  Again, my aim is to develop in parallel, not overwrite the existing xwax code.

ywwg commented:
if run-time configuration is possible, even better.

ywwg commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -604,24 +604,8 @@ void VinylControlXwax::analyzeSamples(CSAMPLE* pSamples, size_t nFrames) {
         }
 
         if (uiUpdateTime(filePosition)) {
-            double pitch_difference = averagePitch - m_dDisplayPitch;
```
if m_dDisplayPitch = dVinylPitch, do we even still need m_dDisplayPitch at all?

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
-struct pitch {
-    double dt, x, v;
+    /* state */
+    double x;      /* position (relative; we subtract dx each step) */
```
Can you also document the physical unit for all values below.

jclsn replied:
There is not really a physical unit here. The velocity is the pitch ranging from being a signed double, where 0.0 is no movement, 1.0 steady playback and -1.0 reverse playback. The position is the current value of the carrier wave. Units don't really help here to understand the filter.

JoergAtGithub replied:
> The velocity is the pitch ranging from being a signed double, where 0.0 is no movement, 1.0 steady playback and -1.0 reverse playback.

This sounds like what we call "rate" at other places of the Mixxx code. Does XWAX uses the term "velocity" for this?

> The position is the current value of the carrier wave.

This explanation is not clear to me, is the position is in 2*Pi or in samples, or the amplitude of the sine wave at the current position?

JoergAtGithub replied:
If XWax uses this notation, than you should keep it!

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+    /* covariance P (2x2, symmetric) */
+    double Pxx, Pxv, Pvv;
+
+    /* Residual treshold for the mode switch */
```
"Residual treshold" does not help me. Is it a common term? Can we add more info?

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+
+    /* Use move semantics for efficiency */
+    p->q  = &p->q_reactive;
+    p->r  = &p->r_reactive;
```
Extra space before =

jclsn replied:
Removed

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+     *            dt^2/2  dt]
+     */
+
+    const double F11 = 1.0, F12 = dt;
```
Is better readable IMHO 
```suggestion
    const double F11 = 1.0;
    const double F12 = dt;
```

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
- * WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
- * General Public License for more details.
+ * Kalman filter replacement for alpha-beta pitch estimator
```
This comment is no longer valid after merging. 
Maybe you can reference the version number up to the old filter was uses it such.

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+ *   v: velocity (pitch)
  *
+ * Observation:
+ *   z = dx  (measured displacement over dt)
```
For my understanding the Kalman filter has a model of the vinyl movements. All measurements are argumenting this model and the the output is read from the plausiblised model instead directly from the input. 
Is this correct? 

Can you add a brief description about these assumptions? Often if is obvious when writing the code but puzzling after years.

jclsn replied:
This is a simple one-dimensional Kalman filter with a position-velocity model. It's the simplest filter that is out there. I can try to document the algorithm better

daschuer commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+    /* Residual treshold for the mode switch */
+    double mode_threshold;
+
+    /* Stable and reactive parameters for the mode switch */
```
Stable is for normal running, reactive for scratching, right? Can you add a comment about it? 
What category are nudging actions to beat match?

daschuer reviewed (commented):
Great that this works better. I have some old knowledge of a kalman filter, where two independent inputs are used. But I must admit I could not fully wrap my head around it. 

Hopefully some comments will help. I added some review comments. 

Do we have here dependent inputs from an earlier filter stage?

JoergAtGithub commented:
As this is fixes an important bug in our DVS system, this should go into 2.5.3 !

Swiftb0y commented:
Is improving the filtering really a bugfix, or is this just because Phil Morse insists on testing 2.5.3 instead of `main`?

ywwg commented:
> Is improving the filtering really a bugfix, or is this just because Phil Morse insists on testing 2.5.3 instead of `main`?

If we can prove that there is no new risk of crashing (no new variables on the heap, for instance), and no regression in performance, then I'd be ok merging this to 2.5.  If there are too many questions then I am ok merging this to main instead.  If Phil complains about sticker drift etc, we can tell him we're working on it but it's not fit for release yet. Personally I think it's worth taking the risk to give him a version that puts the best light on mixxx.

Swiftb0y commented:
> It fixes sticker drift, one of the most reported bugs reported about our DVS system.

Unfortunately, as already pointed out, it doesn't. Its about the same as far as I can tell from my own experiments. I think a better solution to combating sticker drift is a solution I already pointed out to @jclsn in a Zulip DM earlier: Instead of throwing away the position information we get from the LFSR in Relative mode, we should use it periodically to correct for sticker drift. In that sense, it would be more like the absolute mode but with configurable offset from the start.

Swiftb0y reviewed (commented):
couple more thoughts

Swiftb0y commented:
>  no regression in performance

Thats difficult to judge without a benchmark, based on the code though, the algorithmic complexity is the same, so I wouldn't expect the performance to differ by huge amounts (compared to the CPU usage taken up by the rest of mixxx)

ArthurVimond commented:
Hi everyone!
I made some testing with this PR and I did notice some sticker drift, in fact, even more than with version 2.5.2 (which now compensates the drift in relative mode).

I'm not questioning the use of Kalman filter, of course, but just to draw your attention that there is still some sticker drift, at least with my setup. Maybe it's just some values adjustment to find.

I'm on macOS Silicon, using a M-Audio Connectiv sound card, Serato vinyls, Stanton ST-150 turntable and Shure M44-7 cardridge.

ywwg commented:
> I'm heavily scratching on the same spot

what latency are you using?

Also, there was a bug where the timecode would not be initialized at the correct sample rate. Please rebuild with https://github.com/mixxxdj/mixxx/pull/15195 and see if that helps

Swiftb0y commented:
> what latency are you using?

Depends, for live usage usually ~20ms to avoid underflows when using a bunch of decks with rubberband. For testing scratching specifically I can go down to ~2.5ms. The 

> Also, there was a bug where the timecode would not be initialized at the correct sample rate. Please rebuild with https://github.com/mixxxdj/mixxx/pull/15195 and see if that helps

I'm aware, but I'm running at 41.1kHz universally, so this couldn't have been the issue.

Swiftb0y commented:
The offset problems may also just come from my semi jank setup though. The turntable platter I'm using is kinda wobbly, the tonearm is prone to sliding outwards (even though the anti-skate is at 0), the table isn't very stable, etc.

ywwg commented:
With 5ms latency + good turntable + no rubberband, I get no sticker drift, and I really tried!  I am not sure we can fix it in 100% of cases, but any amount of noise will cause the pitch signal to be slightly inaccurate.

daschuer commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -604,24 +604,8 @@ void VinylControlXwax::analyzeSamples(CSAMPLE* pSamples, size_t nFrames) {
         }
 
         if (uiUpdateTime(filePosition)) {
```
I see. But it looks like it is broken: 
```
bool VinylControlXwax::uiUpdateTime(double now) {
    if (m_dUiUpdateTime > now || now - m_dUiUpdateTime > 0.05) {
        m_dUiUpdateTime = now;
        return true;
    }
    return false;
}
```

It uses a magic 0.05 constants with unclear unit. It is actually the file filePosition in s, which pitch dependant, not "now" which implies wall clock time. 

I suggest to resample the value with a constant interval of 50 ms instead and add an appropriate smoothing filter to avoid aliasing noise.

ywwg replied:
yes this is out of scope for this PR.  if the UI update rate is a problem, please file an issue and we can fix it in a separate PR.

daschuer commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -604,24 +604,8 @@ void VinylControlXwax::analyzeSamples(CSAMPLE* pSamples, size_t nFrames) {
-                // For extremely small changes, converge very slowly.
-                m_dDisplayPitch += pitch_difference * .01;
-            }
+            m_dDisplayPitch = dVinylPitch;
```
m_dDisplayPitch can now be disposed.

jclsn replied:
Removed

jclsn committed:
xwax: Adjust gain compensation limit

Since the signal loss of the derivative is compensated dynamically, the
amplitude can be too high when the record is not spinning and you see a
small signal. This doesn't fix this completely, but 25.0 is the sweet
spot of reactiveness and sufficient amplification.

jclsn committed:
preferences: Add pitch estimator selection

Adds a selection combobox to the vinylcontrol panel to choose between the
Alpha-Beta or Kalman filter.

jclsn commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+
+    double Pxx_upd = (1.0 - Kx) * Pxx_pred;
+    double Pxv_upd = (1.0 - Kx) * Pxv_pred;       /* equals Pvx_upd */
+    double Pvv_upd = Pvv_pred - Kv * Pxv_pred;
```
Changed

jclsn commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,229 @@
+        perror(__func__);
+        return;
+    }
+
```
Added the check

jclsn commented on lib/xwax/pitch.h:
```diff
@@ -1,71 +1,213 @@
+    }
+
+    p->q = q;
+    p->r = r;
```
Changed

Swiftb0y commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -171,7 +172,15 @@ VinylControlXwax::VinylControlXwax(UserSettingsPointer pConfig, const QString& g
     s_xwaxLUTMutex.lock();
 
-    timecoder_init(&timecoder, tc_def, speed, sampleRate.value(), /* phono */ false);
+    bool use_legacy_pitch_filter;
+
+    if (strPitchEstimator == MIXXX_VINYL_PITCH_FILTER_LEGACY) {
+        use_legacy_pitch_filter = true;
+    } else {
+        use_legacy_pitch_filter = false;
+    }
```
NIT: This is C++, so I'm gonna be stricter about the style complaints here ;)

Just like in plain C (moreso even in C++), use `const` whenever possible.  
Incidentally that also prevents you from the other grave mistake: declaring variables without initializing them (in this case its trivial so see that its always initialized, but future refactorings may make that harder). So avoid declaring variables without defining them (`const` forces you to do that).

```suggestion
    const bool use_legacy_pitch_filter = strPitchEstimator == MIXXX_VINYL_PITCH_FILTER_LEGACY;
```

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,240 @@
+
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
```
not necessary since you removed the `printf`s, right?
```suggestion
```

Swiftb0y reviewed (commented):
couple more comments. This is looking very good. 

(I know this is just a draft PR, but its hard not to look at the code considering how exciting this change is IMO ;)

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,240 @@
+ * Feed one observation: in the last dt seconds, position moved by dx
+ */
+
+static inline void pitch_kalman_update(struct pitch_kalman* p, double dx) {
+
```
Yeah true, it's just he xwax style. It's questionable if inlining does much in the first place. I personally never use it.

Swiftb0y replied:
definitely makes sense for getters/setters (like `pitch_kalman_current`), but not for a function this big... I mean `inline` doesn't tell the compiler to inline (it decides whether to do that on its own), its telling the compiler/linker how to deal with duplicate symbols.

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+    P_pred[x][v] = F[v][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[v][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][v];
+
+    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
+                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
+
```
If you're modelling the entire matrix, then also initialize the symmetric element.


```suggestion
    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
    P_pred[v][x] = P_pred[x][v];
```

jclsn replied:
It's a draft

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+     *   Since Hx = 1, Hv = 0, this reduces to: y = dx - x_pred
+     */
+
+    const double Hx = 1.0, Hv = 0.0;
+    const double y = dx - (Hx * x_pred + Hv * v_pred);
```
Its great to have these in-depth explanations, especially on why terms simplify a a certain way, so lets use these simplifications in the code and keep the explanations verbose in comments only. Since floating point math usually doesn't behave like math in $$\mathbb{R}$$, compiliers usually don't optimize stuff like this unless you turn on `-ffast-math`. https://compiler-explorer.com/z/d9MrdG4rb

```suggestion
    const double y = dx - x_pred;
```

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+     *   P_upd[v][v] = P_pred[v][v] - Kv * P_pred[x][v]
+     */
+
+    double P_upd[2][2];
+
+    P_upd[x][x] = (1.0 - Kx) * P_pred[x][x];
+    P_upd[x][v] = (1.0 - Kx) * P_pred[x][v]; /* equals Pvx_upd */
+    P_upd[v][v] = P_pred[v][v] - Kv * P_pred[x][v];
+
```
same partial initilization issue here IMO.

jclsn replied:
It's a draft

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+
+    /* Indexes for the matrices */
+
+    const size_t x = 0;
+    const size_t v = 1;
```
idk if I'm a fan of these... IMO using the inidicies directly would be easier to read, but I won't insist.

jclsn commented:
> Considering you're already basing off new features on this branch, what are your plans regarding getting this out of draft status?

Going to address the last comments tomorrow. Also want to describe the matrix operations a bit better. Commit messages need to be better as well. Then I'll put it into ready.

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+    const double F[2][2] = { { 1.0,   dt },
+                             { 0.0,  1.0 }};
+
+    const double Q[2][2] = { { p->coeffs->Q * (pow(dt, 3) / 3.0),  p->coeffs->Q * (pow(dt, 2) / 2.0) },
+                             {                               0.0,  p->coeffs->Q * dt                 }};
```
the previous version with `dt2` and `dt3` was IMO easier to read and faster (avoid calling into `cmath` in hot code if its reasonable to avoid).

jclsn replied:
It's a draft

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+    const double Kx = P_pred[x][x] * invS;
+    const double Kv = P_pred[x][v] * invS;
+
+    /*
+     * Update state:
+     *
+     * x_upd = x_pred + Kx * y
+     * v_upd = v_pred + Kv * y
+     */
```
not super useful IMO, rephrase in terms of the more abstract matrix operations this is modelling?

```suggestion
    /*
     * Update state:
     *
     * xv_upd = xv_pred + K * y
     */
```

jclsn replied:
It's a draft

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,325 @@
+     *   I - K H = [ 1-Kx     0
+     *              -Kv       1 ]
+     *
+     * Then:
+     *   P_upd[x][x] = (1-Kx) * P_pred[x][x]
+     *   P_upd[x][v] = (1-Kx) * P_pred[x][v]   (same as Pvx_upd)
+     *   P_upd[v][v] = P_pred[v][v] - Kv * P_pred[x][v]
```
```suggestion
```

jclsn replied:
It's a draft

Swiftb0y reviewed (commented):
couple more (very nitpicky) comments

Swiftb0y reviewed (changes requested):
no nitpicks now. Only significant change requests

ywwg commented on lib/xwax/pitch_kalman.c:
```diff
@@ -0,0 +1,212 @@
+    const double F[2][2] = { { 1.0,   dt },
+                             { 0.0,  1.0 }};
+
+    const double Q[2][2] = { { p->coeffs->Q * (pow(dt, 3) / 3.0),  p->coeffs->Q * (pow(dt, 2) / 2.0) },
+                             {                               0.0,  p->coeffs->Q * dt                 }};
```
I agree, macros are a bit of a holdover from when programmers had to do optimization by themselves.  Compilers now are *almost*  always much better than us, so it's better to write "plain" code and let the compiler figure out what to inline.  (Remember too that CPUs have their own optimization code -- it's not feasible to guess what they are actually going to do).

Concretely, we had some MMX code in mixxx that was pretty good, and when we rewrote that to plain code, the compiler made it even faster. We also had our own zero-denormal code that caused audio artifacts. Plainly put, usually when we try to optimize, absent specific benchmarks proving otherwise, we don't actually make anything better.

I would suggest writing plainly first, and then after that, throw a profiler at it if you want to see if macros or some other optimization actually makes things faster.

jclsn replied:
Changed

jclsn commented on src/preferences/dialog/dlgprefvinyl.cpp:
```diff
@@ -65,6 +65,15 @@ DlgPrefVinyl::DlgPrefVinyl(
         box->addItem(MIXXX_VINYL_SPEED_45);
     }
 
+    m_vcPitchEstimatorBoxes = {ComboBoxPitchEstimator1,
+            ComboBoxPitchEstimator2,
+            ComboBoxPitchEstimator3,
+            ComboBoxPitchEstimator4};
+    for (auto* box : std::as_const(m_vcPitchEstimatorBoxes)) {
+        box->addItem(MIXXX_VINYL_PITCH_FILTER_KALMAN);
+        box->addItem(MIXXX_VINYL_PITCH_FILTER_LEGACY);
+    }
```
@ywwg What's your take on this?

ywwg replied:
The usage you describe (comparing filters) is valid for a developer use-case, so we can leave it for now. But for the public release I would prefer to either pare down or remove this option entirely.  Adding too many options can overwhelm users, so there is a non-zero cost to adding more options. Mixxx already has a reputation as being difficult to set up, and I want to actively improve that reputation. Unless there is a **proven** performance / quality issue with the new filter that might lead someone to use the legacy filter, once testing is complete I do not see the value of having an option.

Swiftb0y reviewed (approved):
LGTM. Giving the others a chance to comment too. Considering we already had an issue from too many nitpicky comments from my side, I want to emphasize that any additional comments should also be limited to those strictly necessary.

daschuer commented on lib/xwax/timecoder.c:
```diff
@@ -435,7 +437,19 @@ void timecoder_init(struct timecoder *tc, struct timecode_def *def,
+                KALMAN_COEFFS(1e-4, 1e-1), /* medium mode */
+                KALMAN_COEFFS(1e-1, 1e-4), /* reactive mode */
+                6e-4,  /* medium threshold  */
+                15e-4); /* scratch threshold  */
```
Is this from 16 x speed?

daschuer reviewed (changes requested):
For my understanding some of the filter constants are not dt compensated. Did you test scratching with 96 kHz? 

Thank you for the overall very good description. I have stumbled about some minor issues. See inline comments.

daschuer commented on lib/xwax/pitch_kalman.c:
```diff
@@ -0,0 +1,225 @@
+
+enum { x = 0, v = 1 }; /* Matrix indices for readability */
+
+static inline double pow2(double x)
+{
+    return (x * x);
+}
+
+static inline double pow3(double x)
+{
+    return (x * x * x);
+}
```
here it is: 

```suggestion
static inline double pow2(double val)
{
    return (val * val);
}

static inline double pow3(double val)
{
    return (val * val * val);
}
```

jclsn replied:
Approved

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,103 @@
+ *   v : Velocity (pitch)
+ *
+ * Assumption: :
+ *   dx: We advanced a quarter of the cycle
```
No, it is not a revolution of the vinyl. A revolution on the unit circle (the scope), which corresponds to four zeros, similar to this graphic

https://en.wikipedia.org/wiki/In-phase_and_quadrature_components#/media/File:Phase_shifter_using_IQ_modulator.gif

A zero crossing is when the phasor passes one of the axes.

jclsn replied:
Added more comments

Swiftb0y commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,103 @@
+        return 0.0;
+    }
+
+    return p->Xk[1];
```
explicitly changed. see https://github.com/mixxxdj/mixxx/pull/15194#discussion_r2283373374

jclsn replied:
I just added this back from v to 1 upon Niko's request, because else the `x` and `v` in the head would pollute all other source files. What I can do is move this function into the `pitch_kalman.c` though.

jclsn replied:
Added a comment

jclsn commented on lib/xwax/pitch_kalman.c:
```diff
@@ -0,0 +1,225 @@
+
+    p->P[x][x] = 1e6;
+    p->P[x][v] = 0.0;
+    p->P[v][v] = 1e6;
```
Added a comment

daschuer commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,106 @@
+        return 0.0;
+    }
+
+    /* Return the velocity Xk[v] */
```
Xk[v] / tc->speed is the rate ratio which is 1 at unity. tc->speed is also 1 for a 33 r/min vinyl. 

See: 
https://github.com/mixxxdj/mixxx/blob/ba4a0f9e11aefe0bc9f1ce6328da26ed89840adf/src/vinylcontrol/vinylcontrolxwax.cpp#L144

```suggestion
    /* Return the velocity Xk[v] relative to 33.3 r/min */
```

jclsn replied:
Improved the comment

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,106 @@
+ *   x : Position on the unit circle or complex plane (see the scope)
+ *   v : Velocity (pitch)
+ *
+ * Assumption: We steadily advance with a constant velocity
+ *   dx: Usually the displacement is 1/4 of cycle of the sinusoid π/2
+ *       or one zero crossing
```
This is not an accurate description of what is happening. See https://github.com/mixxxdj/mixxx/pull/15194#discussion_r2286198341

jclsn replied:
Added an extended explanation in the comment.

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,106 @@
+ *
+ * State:
+ *   x : Position on the unit circle or complex plane (see the scope)
+ *   v : Velocity (pitch)
```
Improved the comment

jclsn commented:
> For my understanding some of the filter constants are not dt compensated. Did you test scratching with 96 kHz?
> 
> Thank you for the overall very good description. I have stumbled about some minor issues. See inline comments.

Of course the sampling frequency is used in the filter. 

https://github.com/mixxxdj/mixxx/blob/6da2e9ec7624c825b39a0adf54c931c0243d800f/lib/xwax/pitch_kalman.c#L37

daschuer commented on lib/xwax/timecoder.c:
```diff
@@ -435,7 +437,19 @@ void timecoder_init(struct timecoder *tc, struct timecode_def *def,
+                KALMAN_COEFFS(1e-8, 10.0), /* stable mode */
+                KALMAN_COEFFS(1e-4, 1e-1), /* medium mode */
+                KALMAN_COEFFS(1e-1, 1e-4), /* reactive mode */
+                6e-4,   /* medium threshold  */
```
The threshold is compared with the Inovation, the diffrence of the predicted x with the new x.  
`const double y = dx - X_pred[x];` 

Unfortunately this is a fast changing value: 

dx is like that at 44100 Hz  
0.0
0.0
0.0 
0.0 
0.00025
0.0
0.0
0.0
0.0
0.0
0.0

X_pred[x] is like 0,000025 (or -) I am still struggling here ...

At 96 kHz we have still 0.00025 but double as much zeros X_pred[x] has only halve the value.  
This means that we still a samplerate dependency here, because in most cases "y = -X_pred[x]"

```suggestion
                6e-4,   /* medium threshold  */
```

jclsn replied:
The filter knows about the samplerate as $dt = \frac{1}{samplerate}$. Everything works well. Please try this out yourself before making claims like this!

daschuer reviewed (commented):
I hope I have now a better understanding how the values are scaled. See my comments.

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,112 @@
+        return 0.0;
+    }
+
+    /* Return the velocity Xk[v] relative to carrier frequency of the timecode */
```
It is still relative to the carrier frequency. It is just normalized. I added this to the comments.

jclsn commented on lib/xwax/pitch_kalman.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * Assumption: We steadily advance with a constant velocity
+ *
+ *   dx: The displacement of the position by which we advanced. When
```
I improved the comments.

jclsn commented:
@daschuer I added a comment explaining the noisiness of the innovation. Is it okay like this?

I also added a small fix that sets the filter back into reactive mode when falling under a certain pitch. I could successfully fix the ghost pitch issue with this :)

The PR should be ready now!

daschuer commented:
Looks good. Thank you. You can squash the fixup and then we can merge.

jclsn commented:
Pushed some small formatting improvements. Good to merge!

Swiftb0y commented:
Just tested it, still works wonderfully. @daschuer please give your formal LGTM.

daschuer reviewed (approved):
LGTM, thank you for taking the time for this nice improvement und for the effortsvto explain it that even I understood it :-)

JoergAtGithub commented:
Yes, thats was the common sense in the last core dev meeting.

Swiftb0y commented:
I did ;) https://github.com/mixxxdj/website/pull/398#discussion_r2274178926 You just need to cherry-pick the newest commit again.
